sips/sip-025/sip-025-execution-plan.md

```diff
@@ -0,0 +1,144 @@
+
+This SIP reevaluates the existing approach and proposes a new design that gives developers more freedom in assembling their execution plan (a set of dynamic contract IDs to invoke), while also providing users with a method to verify, after execution, that the executed plan matches with what they committed to.
+
+Consequently, if the execution plan changes between the time a user signs and broadcasts their transaction and the time the transaction is executed, the said transaction would be rolled back, similar to a rollback triggered by a post-condition failure.
+
+The concept of Optimistic Dynamic Dispatch is defined as follows: by default, the smart contract is trusted to correctly route the contract calls. Additionally, a mechanism is provided for users to have the transaction rolled back if the execution ends up incorporating components that were not initially specified.
+
+# **License and Copyright**
+
+This SIP is made available under the terms of the Creative Commons CC0 1.0 Universal license, available at https://creativecommons.org/publicdomain/zero/1.0/ 
+
+This SIP's copyright is held by the Stacks Open Internet Foundation.
+
+# **Specification**
+
+The novel post conditions mechanism introduced in Stacks 2.0 is being utilized to design optimistic dynamic dispatch. 
+
+## Execution Plan Commitment Usage
+
+### Augmenting Transaction Wire Format
+
+Users can require an execution plan commitment in the post conditions:
+
+```markdown
+A 1-byte post-condition mode, identifying whether or not post-conditions must fully cover all transferred assets. It can take the following values:
+- 0x01 (existing): This transaction may affect other assets not listed in the post-conditions.
+- 0x02 (existing): This transaction may NOT affect other assets besides those listed in the post-conditions.
+- 0x03 (new): This transaction may follow a constrained execution plan.
+```
+
+### Augmenting Transaction Receipts
+
+During the execution of a contract-call transaction, the Clarity VM builds a tree of the different contracts and public methods invoked.
+
+All the branches of this tree are collected, hashed and added in their order of execution in a Merkle Tree. The Merkle Root constitute the Execution Plan Commitment that users can incorporate in their transactions post conditions.
+
+### Augmenting HTTP Endpoint `/v2/contracts/call-read`
+
+In order to help users computing an execution plan commitment, the JSON response of the existing HTTP endpoint `/v2/contracts/call-read/` can be augmented to return:
+
+```json
+{
+  "okay":true,
+	"execution_plan_commitment": "0x607b5e64b8d51f78ac5a37984e64f6493e9a7f90b605fa068be86ff035f48141"
+}
+```
+
+This HTTP endpoint must also be able to execute any kind of function `readonly` , but also `readwrite` simulating writes to an in-memory store.
```

Comment:
for added clarity what about:
...of function — `readonly` as well as `readwrite` — simulating writes to an in-memory store.